### DISCUSSION TITLE
How to define a complex const in WGSL?

### DISCUSSION THREAD

**Discussion by Liburia** [gpuweb/gpuweb#3575] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/2414193/200164824-567b731c-5d32-491d-adc7-fecdf01719a4.png)

**Comment by ben-clayton** [comment, marked as answer]
There's a distinction between a `const` variable (shader-creation-time evaluated + immutable) which can be initialized with a `const expression` and a `let` variable (runtime evaluated + immutable) which can be initialized with any expression. I'm not entirely sure which one you want here.

If you want a true shader-creation-time (compile-time) evaluated constant, use `const` for these variable declarations. These can be declared at both module-scope (global) and function scope (local). Be aware that builtin constant expression evaluation is still being implemented in tint (Chromium), so you may find some builtins trigger errors when they shouldn't. I don't know how far along Naga (wgpu / Firefox) is with their implementation here.

If you're after an immutable, runtime-evaluated variable, then you want `let`, but these can only be declared at function scope.

`<speculation>` Based on the error, I suspect this error message came from Naga or another compiler, and may be due to these `let` declarations being module-scope? If that's correct, then the compiler should be saying that you cannot use a `let` declaration at module-scope`</speculation>`

**Reply by Liburia** [reply]
I am using cargo-wgsl, which, as far as I understand uses naga.

Looks like it's an [issue there.](https://github.com/gfx-rs/naga/issues/2071)
